Fix bipartite graph constructor for reduced adjacency matrix with `immutable=True`

zezoo050 opened the pull request:
Related to issue `#39295` 

Previously, when a reduced adjacency matrix was provided and `immutable=True` was set, the graph construction failed. This fix ensures that the graph is correctly constructed and behaves as expected.

This issue occurred because the parent class `Graph` constructor was called with `immutable=True` and then attempted to add vertices and edges after construction, which is not allowed.

The fix prepares the vertex list and edge list before calling the parent class constructor with the precomputed vertices and edges.


### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

zezoo050 committed:
Fix bipartite graph constructor for reduced adjacency matrix with immutable=True

Previously, when a reduced adjacency matrix was provided and `immutable=True`
was set, the graph construction failed. This fix
ensures that the graph is correctly constructed and behaves as expected.

dcoudert commented on src/sage/graphs/bipartite_graph.py:
```diff
@@ -467,32 +467,34 @@ def __init__(self, data=None, partition=None, check=True, hash_labels=None, *arg
-            self.add_vertices(self.left)
-            self.add_vertices(self.right)
-
+            vertices = list(self.left) + list(self.right)
```
this is not needed. You can use directly `range(nrows + ncols)` in the call to `Graph.__init__(...)`

dcoudert commented on src/sage/graphs/bipartite_graph.py:
```diff
@@ -467,32 +467,34 @@ def __init__(self, data=None, partition=None, check=True, hash_labels=None, *arg
+
+            # ensure that construction works
+            # when immutable=True (trac #39295)
+            Graph.__init__(self, data=[vertices, edges], format='vertices_and_edges')
```
other parameters in `*args` and `**kwds` are not taken into account.

dcoudert reviewed (commented):
Thank you for this PR.

You must also add a doctest showing that the issue is fixed.

zezoo050 committed:
Fix BipartiteGraph construction with immutable=True

- Use `range(nrows + ncols)` in `Graph.__init__`.
- Correctly call `Graph.__init__` with `*args` and `**kwds`.
- Added a doctest.

dcoudert reviewed (approved):
This fixes the issue.

I tested on my laptop as I can't launch CI tests.

user202729 commented on src/sage/graphs/bipartite_graph.py:
```diff
@@ -467,32 +477,31 @@ def __init__(self, data=None, partition=None, check=True, hash_labels=None, *arg
+                            edges.append((ii, jj + ncols))
+
+            # ensure that construction works
+            # when immutable=True (trac #39295)
```
Use `issue` instead of `trac`.

By the way is there something in the documentation that suggests using `trac`? If so we should modify it.

zezoo050 replied:
Changed it to issue.

I didn't find any instruction in the documentation for writing comments about the fixes. I only found it while reading some of the source code, so I used it.

zezoo050 committed:
Fix the comment of the issue number

zezoo050 committed:
Replaced edges list with a generator

dcoudert reviewed (approved):
LGTM.

tested on my laptop as I cannot launch the CIs.
